codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -163,67 +178,120 @@ private static int compressionLevel(int blockSize) {
+        }
+
+        // in addition to just the raw byte count, the headers (HEADER_LENGTH) per block (configured via
+        // #blockSize) will also add to the targetBufSize, and the combination of those would never wrap around
+        // again to be >= 0, this is a good check for the overflow case.
+        if (targetBufSize > maxEncodeSize || 0 > targetBufSize) {
+            String errorMsg = "requested encode buffer size (%d bytes) exceeds the maximum allowable size (%d bytes)";
+            throw new EncoderException(String.format(errorMsg, targetBufSize, maxEncodeSize));
+        }
+
+        if (allowEmptyReturn && targetBufSize < blockSize) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        if (preferDirect) {
+            return ctx.alloc().ioBuffer(targetBufSize, targetBufSize);
+        } else {
+            return ctx.alloc().heapBuffer(targetBufSize, targetBufSize);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Encodes the input buffer into {@link #blockSize} chunks in the output buffer. Data is only compressed and
+     * written once we hit the {@link #blockSize}; else, it is copied into the backing {@link #buffer} to await
+     * more data.
+     */
+    @Override
     protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws Exception {
         if (finished) {
             out.writeBytes(in);
             return;
         }
 
-        int length = in.readableBytes();
-
         final ByteBuf buffer = this.buffer;
-        final int blockSize = buffer.capacity();
-        while (currentBlockLength + length >= blockSize) {
-            final int tail = blockSize - currentBlockLength;
-            in.getBytes(in.readerIndex(), buffer, currentBlockLength, tail);
-            currentBlockLength = blockSize;
-            flushBufferedData(out);
-            in.skipBytes(tail);
-            length -= tail;
+        int length;
+        while ((length = in.readableBytes()) > 0) {
+            final int nextChunkSize = Math.min(length, buffer.writableBytes());
+            in.readBytes(buffer, nextChunkSize);
+
+            if (!buffer.isWritable()) {
+                flushBufferedData(out);
+            }
         }
-        in.readBytes(buffer, currentBlockLength, length);
-        currentBlockLength += length;
     }
 
     private void flushBufferedData(ByteBuf out) {
-        int currentBlockLength = this.currentBlockLength;
-        if (currentBlockLength == 0) {
+        int flushableBytes = buffer.readableBytes();
+        if (flushableBytes == 0) {
             return;
         }
         checksum.reset();
-        checksum.update(buffer.array(), buffer.arrayOffset(), currentBlockLength);
+        checksum.update(buffer, buffer.arrayOffset(), flushableBytes);
         final int check = (int) checksum.getValue();
 
-        out.ensureWritable(compressedBlockSize);
+        final int bufSize = compressor.maxCompressedLength(buffer.readableBytes()) + HEADER_LENGTH;
```

Comment:
`buffer.readableBytes()` -> `flushableBytes` (re-use since you have already calculated readable bytes).